### PR TITLE
F leader reconcile summaries

### PR DESCRIPTION
**Pull request by diptanu**

### CODE REVIEW

File: nomad/fsm.go [hunk 1]
```diff
@@ -448,6 +448,7 @@ func (n *nomadFSM) applyAllocClientUpdate(buf []byte, index uint64) interface{}
 
 // applyReconcileSummaries reconciles summaries for all the jobs
 func (n *nomadFSM) applyReconcileSummaries(buf []byte, index uint64) interface{} {
+	n.logger.Printf("[DEBUG] fsm: reconciling job summaries at index: %v", index)
```
Comment (dadgar) [inline nomad/fsm.go]: Remove

File: nomad/fsm.go [hunk 2]
```diff
@@ -604,6 +605,7 @@ func (n *nomadFSM) Restore(old io.ReadCloser) error {
 	// If the index is 0 that means there is no job summary in the snapshot so
 	// we will have to create them
 	if index == 0 {
+		n.logger.Printf("[DEBUG] fsm: reconciling job summaries during state restore at index: %v", index)
```
Comment (dadgar) [inline nomad/fsm.go]: Remove

File: nomad/leader.go [hunk 3]
```diff
@@ -166,6 +167,17 @@ func (s *Server) establishLeadership(stopCh chan struct{}) error {
 		s.logger.Printf("[ERR] nomad: heartbeat timer setup failed: %v", err)
 		return err
 	}
+
+	// COMPAT 0.4 - 0.4.1
+	// Reconcile the summaries of the registered jobs. We reconcile summaries
+	// only if the server is 0.4.1 since summaries are not present in 0.4 they
+	// might be incorrect after upgrading to 0.4.1 the summaries might not be
+	// correct
+	if strings.HasPrefix(s.config.Build, "0.4.1") {
```
Comment (dadgar) [inline nomad/leader.go]: I don't think we can have this check? What happens if someone goes from 0.4 -> 0.5

File: nomad/leader.go [hunk 4]
```diff
@@ -458,6 +470,25 @@ func (s *Server) reconcileMember(member serf.Member) error {
 	return nil
 }
 
+// reconcileJobSummaries reconciles the summaries of all the jobs registered in
+// the system
+// COMPAT 0.4 -> 0.4.1
+func (s *Server) reconcileJobSummaries() error {
+	index, err := s.fsm.state.LatestIndex()
```
Comment (dadgar) [inline nomad/leader.go]: Can remove 477-482

File: nomad/leader.go (hunk 3 above)
Comment (diptanu) [inline nomad/leader.go]: We can update this check when we do 0.5.0 and check if the version is either 0.4.1 or 0.5.0

File: nomad/leader.go (hunk 4 above)
Comment (diptanu) [inline nomad/leader.go]: The debug message below would indicate the new leader has reconciled the summaries. 

There is no way to differentiate between reconciliation by user and leader by just looking at fsm logs.

File: nomad/leader.go (hunk 3 above)
Comment (dadgar) [inline nomad/leader.go]: That is the same as removing the check?

Comment (diptanu) [inline nomad/leader.go]: Agreed. I will remove it.

**Pull request merged**